### PR TITLE
Add model initialization for CustomSpec

### PR DESCRIPTION
**Pull request by krabhishek8260**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Support for #219

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/222)

### CODE REVIEW

**Comment by ellis-bigelow** [comment]
/hold
Hey Abhishek, interesting approach.

I think "KFSERVING_CUSTOMSPEC_MODEL_URI" may end up being a little bit controversial. I'd love to see a little writeup including:

1. What your PR does
2. What problem you're trying to solve
3. Alternatives you've considered and why you picked this design choice. 

That will be a great place to start the discussion.

**Comment by rakelkar** [comment]
@krabhishek8260 lets work together to contrast this against other possible approaches.

**Comment by rakelkar** [comment]
let's use https://github.com/kubeflow/kfserving/issues/33 to have this discussion

File: pkg/webhook/admission/deployment/model_initializer_injector.go [hunk 1]
```diff
@@ -137,6 +137,13 @@ func (mi *ModelInitializerInjector) InjectModelInitializer(deployment *appsv1.De
 	}
 	userContainer.VolumeMounts = append(userContainer.VolumeMounts, sharedVolumeReadMount)
 
+	// Change the CustomSpecModelUri env variable value to the default model path if present
+	for index, envVar := range userContainer.Env {
+		if envVar.Name == constants.CustomSpecModelUriEnvVarKey && envVar.Value != "" {
+			userContainer.Env[index].Value = constants.DefaultModelLocalMountPath
+		}
+	}
```
Comment (hougangliu) [inline pkg/webhook/admission/deployment/model_initializer_injector.go]: I have no idea why L140-L145 is in need. It seems they can be dropped. Please correct me if you don't agree with me

Comment (krabhishek8260) [reply pkg/webhook/admission/deployment/model_initializer_injector.go]: The user relies on this env variable in his custom script to load the models from the mount location.

Comment (hougangliu) [reply pkg/webhook/admission/deployment/model_initializer_injector.go]: OK, SGTM! thanks!

Comment (ellis-bigelow) [reply pkg/webhook/admission/deployment/model_initializer_injector.go]: Magic environment variables don't SGTM ;)

**Comment by ellis-bigelow** [comment]
As discussed in person last Friday, let's open up the conversation for adding a StorageUri to the ModelSpec and allowing a STORAGE_URI environment variable to be passed through the custom spec.

**Comment by ellis-bigelow** [comment]
Let's use storageuri in new place and update the old ones in alpha2

File: pkg/webhook/admission/deployment/model_initializer_injector.go (hunk 1 above)
Comment (animeshsingh) [reply pkg/webhook/admission/deployment/model_initializer_injector.go]: Good to see this optional

**Comment by rakelkar** [comment]
How would StorageUri work for whitebox explainers?

**Comment by cliveseldon** [comment]
* Blackbox explainers (Alibi for now) need access to a predictURI. This is available as kfserving creates that URI and it can be injected. 
 * Whitebox explainers (none for now) need access to the saved model to load in a known format, e.g. Tensorflow whitebox explainer for Tensorflow predictor. That is available in the yaml in `modelUri` so can be injected and downloaded via modelInitializer logic.

Both could also have saved pickled versions of themselves which could use the modelInitializer logic to download them to a known location.

**Comment by ellis-bigelow** [comment]
I think this is basically ready to go,
@yuzisun @cliveseldon @animeshsingh @rakelkar, do you have any thoughts?

**Comment by animeshsingh** [comment]
So can we add an example in docs/samples folder? Also why are we calling it storage URI? How is it different from model uri field which also points to storage?

**Comment by rakelkar** [comment]
@animeshsingh They are the same thing. The idea is that ModelUri will be renamed to StorageUri in alpha2

**Comment by ellis-bigelow** [comment]
Excited to see this move forward!
/lgtm
/approve

**Pull request merged**